charts/consul/templates/gateway-resources-configmap.yaml

```diff
@@ -60,7 +60,10 @@ data:
             {{- with .Values.meshGateway.nodeSelector }}
             nodeSelector: {{ fromYaml . | toJson }}
             {{- end }}
-            priorityClassName: {{ toJson .Values.meshGateway.priorityClassName }}
+            {{- if .Values.meshGateway.topologySpreadConstraints }}
+            topologySpreadConstraints:
+              {{ tpl .Values.meshGateway.topologySpreadConstraints . | nindent 16 | trim }}
```

Comment:
we should be able to use `{ fromYamlArray .Values.meshGateway.topologySpreadConstraints | toJson}` here and not need to worry about the indentation

Reply:
I'll test it out

Reply:
running it with this resulted in the following error 

``` ["error converting YAML to JSON: yaml: line 1: did not find expected '-' indicator"]```

Reply:
Oh is this not a yaml array? In that case it should be `fromYaml` rather than `fromYamlArray`

Reply:
its a multiline yaml string. I'll try the fromYaml now as well

Reply:
Same error using fromYaml. This is what the field looks like. 

```
  topologySpreadConstraints: |
    - maxSkew: 1
    topologyKey: zone
    whenUnsatisfiable: DoNotSchedule
    labelSelector:
      matchLabels:
        foo: bar
```

Reply:
I think the issue is the format of your YAML string. It should be:

```yaml
- maxSkew: 1
  topologyKey: zone
  whenUnsatisfiable: DoNotSchedule
  labelSelector:
    matchLabels:
      foo: bar
```

Where the subsequent lines after `-` are indented to match the key.

Try using the `fromYamlArray` like this:

```yaml
            {{- with .Values.meshGateway.topologySpreadConstraints }}
            topologySpreadConstraints: {{ fromYamlArray . | toJson }}
            {{- end }}
```

Reply:
Gotcha! I'll test it out now. I just copied the example of the topologySpreadConstraints from the documentation.

Reply:
That still didn't work for me. I'll go ahead and merge, but maybe we can pair on this after the holiday?

Reply:
so I ran this locally with the following values file at `consul-k8s/charts/consul/consul_values.yaml`:
```global:
  name: consul
  datacenter: dc1

  experiments:
    - resource-apis

  image: hashicorppreview/consul:1.18-dev
  imageK8S: consul-k8s-control-plane-dev:local
  imageConsulDataplane: hashicorppreview/consul-dataplane:1.4-dev

connectInject:
  enabled: true

meshGateway:
  enabled: true
  hostNetwork: true
  topologySpreadConstraints: |
    - maxSkew: 1
      topologyKey: zone
      whenUnsatisfiable: DoNotSchedule
      labelSelector:
        matchLabels:
        foo: bar
  nodeSelector: |
    beta.kuberenetes.io/arch: amd64
    beta.kubernetes.io/another: linux
  tolerations: |
    - key: "key1"
      operator: "Equal"
      value: "value1"
      effect: "NoSchedule"
ui:
  enabled: false
```
with the following in the helm chart:
```
            {{- with .Values.meshGateway.topologySpreadConstraints }}
            topologySpreadConstraints:
              {{ fromYamlArray . | toJson }}
            {{- end }}
```

and got the following yaml generated when running `helm template -f ./consul_values.yaml -s templates/gateway-resources-configmap.yaml .` from the `consul-k8s/charts/consul` directory

```
# Source: consul/templates/gateway-resources-configmap.yaml
# Configuration of Gateway Resources Job which creates managed Gateway configuration.
apiVersion: v1
kind: ConfigMap
metadata:
  name: consul-gateway-resources-config
  namespace: default
  labels:
    app: consul
    chart: consul-helm
    heritage: Helm
    release: release-name
    component: gateway-resources
data:
  resources.json: |
    {"limits":{"cpu":"100m","memory":"100Mi"},"requests":{"cpu":"100m","memory":"100Mi"}}
  config.yaml: |
    gatewayClassConfigs:
      - apiVersion: mesh.consul.hashicorp.com/v2beta1
        metadata:
          name: consul-mesh-gateway
        kind: GatewayClassConfig
        spec:
          labels:
            set:
              app: consul
              chart: consul-helm
              heritage: Helm
              release: release-name
              component: mesh-gateway
          deployment:
            container:
              consul:
                logging:
                  level: info
              portModifier: 8000
              resources: {"limits":{"cpu":"100m","memory":"100Mi"},"requests":{"cpu":"100m","memory":"100Mi"}}
            initContainer:
              consul:
                logging:
                  level: info
              resources: {"limits":{"cpu":"50m","memory":"50Mi"},"requests":{"cpu":"50m","memory":"50Mi"}}
            nodeSelector: {"beta.kuberenetes.io/arch":"amd64","beta.kubernetes.io/another":"linux"}
            hostNetwork: true
            priorityClassName: ""
            topologySpreadConstraints:
              [{"labelSelector":{"foo":"bar","matchLabels":null},"maxSkew":1,"topologyKey":"zone","whenUnsatisfiable":"DoNotSchedule"}]
            replicas:
              default: 1
              min: 1
              max: 1
            tolerations: [{"effect":"NoSchedule","key":"key1","operator":"Equal","value":"value1"}]
          service:
            type: LoadBalancer
    meshGateways:
      - apiVersion: mesh.consul.hashicorp.com/v2beta1
        kind: MeshGateway
        metadata:
          name: mesh-gateway
          namespace: default
        spec:
          gatewayClassName: consul-mesh-gateway
```
 what errors were you seeing?